File: myGym/myGym/SessionsViewController.m
//
//  SessionsViewController.m
//  myGym
//
//  Created by SQLI51109 on 24/11/2015.
//  Copyright © 2015 julien gimenez. All rights reserved.
//

#import "SessionsViewController.h"
#import "SessionModel.h"
#import "FSCalendar.h"
#import "DateManager.h"
#import "DayViewcontroller.h"

static NSString * const kSegueDayView = @"showDetailDay";

@interface SessionsViewController ()<FSCalendarDelegate,FSCalendarDataSource>

@property (weak,nonatomic)IBOutlet FSCalendar *myCalendar;
@property (strong,nonatomic) RLMResults *sessionResults;
@property (strong,nonatomic) NSMutableArray<SessionModel*>*dailySession;

@end

@implementation SessionsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _myCalendar.delegate = self;
    _myCalendar.dataSource = self;
    _myCalendar.locale = [[NSLocale alloc] initWithLocaleIdentifier:@"fr_BI"];
    _myCalendar.appearance.headerMinimumDissolvedAlpha = 0.0;
    
    [self getSessions:^(RLMResults *sessions) {
        _sessionResults = sessions;
    }];
}

- (void)getSessions:(void(^)(RLMResults*sessions))completionBlock{
    
    completionBlock([[RealmManager shared]getAllSessions]);
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

// FSCalendarDelegate
- (void)calendar:(FSCalendar *)calendar didSelectDate:(NSDate *)date
{
    _dailySession = [NSMutableArray new];
    
    for (SessionModel *session in _sessionResults) {
        if ([DateManager isSameDayWithDate1:date date2:session.day]) {
            [_dailySession addObject:session];
        }
    }
    
    if ([_dailySession count]) {
        [self performSegueWithIdentifier:kSegueDayView sender:nil];
    }
    else{
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Information" message:@"Aucune séance de prévue" preferredStyle:UIAlertControllerStyleAlert];
        
        UIAlertAction* ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
        [alertController addAction:ok];
        
        [self presentViewController:alertController animated:YES completion:nil];
    }
}

- (void)calendar:(FSCalendar *)calendar didDeselectDate:(NSDate *)date{
    
}

#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
    if([segue.identifier isEqualToString:kSegueDayView]){
        
        DayViewController *dayVC = [segue destinationViewController];
        dayVC.dailySession = _dailySession;
    }
}

@end

File: myGymSwift/myGymSwift/SessionsViewController.swift
//
//  SessionsViewController.swift
//  myGymSwift
//
//  Created by julien gimenez on 28/11/2015.
//  Copyright © 2015 julien gimenez. All rights reserved.
//

import UIKit
import FSCalendar
import JLToast

class SessionsViewController: UIViewController, FSCalendarDelegate, FSCalendarDataSource {

    @IBOutlet weak var myCalendar: FSCalendar?
    let kShowDetailDay = "showDetailDay"
    var selectedDay: String = String()
    var selectedDate: NSDate = NSDate()
    var sessionsArray: Array<SessionObject> = Array<SessionObject>()
    var boolToast: Bool = Bool()
    var timer = NSTimer()
    
    // MARK: - Init

    override func viewDidLoad() {
        super.viewDidLoad()

        myCalendar?.locale = NSLocale.init(localeIdentifier: FormaterManager.SharedInstance.fr_BI)
        myCalendar?.delegate = self
        myCalendar?.dataSource = self
        myCalendar?.appearance.headerMinimumDissolvedAlpha = 0.0
        myCalendar?.appearance.headerDateFormat = FormaterManager.SharedInstance.MMM
    }
    
    // MARK: - FSCalendar delegate
    
    func calendar(calendar: FSCalendar!, didSelectDate date: NSDate!) {

        selectedDay =  FormaterManager.SharedInstance.formatWeekDayAndDate(date)
        selectedDate = date
        
        RealmManager.SharedInstance.isSessionWithDate(date) { (sessions) -> Void in
            
            self.sessionsArray.removeAll()
            
            for session in sessions {
                let sessionObject = SessionObject()
                sessionObject.setSessionForCell(session, completion: { (sessionObject) -> Void in
                    self.sessionsArray.append(sessionObject)
                })
            }
            if(sessions.count > 0) {
                self.performSegueWithIdentifier(self.kShowDetailDay, sender: self)
            }
            else{
                
                if(!self.timer.valid){
                    self.timer = NSTimer.scheduledTimerWithTimeInterval(JLToastDelay.ShortDelay+1.0, target: self, selector: "countUp", userInfo: nil, repeats: false)
                    JLToast.makeText(NSLocalizedString("NOTHING", comment:"")+"\n"+self.selectedDay, duration: JLToastDelay.ShortDelay).show()
                }

                /*let alertController = UIAlertController(title: "", message: NSLocalizedString("NOTHING", comment:""), preferredStyle: .Alert)
                let defaultAction = UIAlertAction(title: NSLocalizedString("CLOSE", comment:""), style: .Default, handler: nil)
                alertController.addAction(defaultAction)
                self.presentViewController(alertController, animated: true, completion: nil)*/
            }
        }
    }
    
    func minimumDateForCalendar(calendar: FSCalendar!) -> NSDate! {
        
        return NSDate()
    }
    
    func maximumDateForCalendar(calendar: FSCalendar!) -> NSDate! {
        
        let date = NSCalendar.currentCalendar().dateByAddingUnit(.WeekOfMonth, value: 1, toDate: NSDate(), options: [])

        return date
    }
    
    // MARK: - Navigation

    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
        if(segue.identifier == kShowDetailDay) {
            
            let svc = segue.destinationViewController as! DayViewController
            svc.sessionsArray = sessionsArray
            svc.selectedDay   = selectedDay
            svc.selectedDate  = selectedDate
        }
    }
    
    func countUp() {
        
        timer.invalidate()
    }
    
    // MARK: - Memory

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

}
